downstairs/src/extent_inner_raw.rs

```diff
@@ -220,24 +225,73 @@ impl ExtentInner for RawInner {
+                first_req.offset.value,
+                n_contiguous_requests as u64,
+            )?;
+            cdt::extent__read__get__contexts__done!(|| {
+                (job_id.0, self.extent_number, n_contiguous_requests as u64)
+            });
+
+            // Now, we'll begin to construct the ReadResponses in-place
+            let mut chunks = vec![];
+            for (ctx, req) in block_contexts
+                .into_iter()
+                .zip(requests[req_run_start..][..n_contiguous_requests].iter())
+            {
+                out.put_u64_le(req.eid);
+                out.put_u64_le(req.offset.value);
+                out.put_u32_le(req.offset.shift);
+
+                // Inner BytesMut, serialized as size + data
+                out.put_u64_le(block_size.into());
+                let data_start = out.len();
+                out.resize(data_start + block_size as usize, 1);
+                check_input(self.extent_size, req.offset, &out[data_start..])?;
+
+                // Vec<BlockContext>
+                if let Some(ctx) = ctx {
+                    out.put_u64_le(1);
+                    bincode::serialize_into(
+                        (&mut out).writer(),
+                        &ctx.block_context,
+                    )
+                    .unwrap();
+                } else {
+                    out.put_u64_le(0);
+                }
+
+                // Split off this chunk, so that we can issue a read to many
+                // mutable chunks in a single syscall
+                let next = out.split_off(out.len());
+                chunks.push(out);
+                out = next;
+            }
+
+            // This is the fun part.  We're going to iterate over our block
+            // contexts, copying them into the output buffer, and also borrowing
+            // chunks of the output buffer
+
             // Create our responses and push them into the output. While we're
             // at it, check for overflows.
-            let resp_run_start = responses.len();
             let mut iovecs = Vec::with_capacity(n_contiguous_requests);
-            for req in requests[req_run_start..][..n_contiguous_requests].iter()
-            {
-                let resp = ReadResponse::from_request(req, block_size as usize);
-                check_input(self.extent_size, req.offset, &resp.data)?;
-                responses.push(resp);
-            }
 
             // Create what amounts to an iovec for each response data buffer.
             let mut expected_bytes = 0;
-            for resp in
-                &mut responses[resp_run_start..][..n_contiguous_requests]
-            {
-                expected_bytes += resp.data.len();
-                iovecs.push(IoSliceMut::new(&mut resp.data[..]));
+            for chunk in &mut chunks {
+                expected_bytes += block_size as usize;
+                iovecs.push(IoSliceMut::new(
+                    &mut chunk[28..][..block_size as usize],
```

Comment:
The 28 here, that's `eid` + `offset.value` + `offset.shift` + `block_size`?
Is there any way to.. hmm.  this is really hand rolled serialization.
Maybe comment what the 28 is for.

If we ever change how this is rolled up, I guess it just has to be done carefully.

Reply:
Yes, that's correct.  I've added a named constant and some internal consistency checks in a184fe0e745f27076ef068cf43bf19a0d8e52c03